Methods do not check errors

Ran into an issue today where an `Interface` type contained a channel, which Go can't marshal into JSON. Obviously this isn't valid, but it didn't occur to me I was marshaling that (my `Interface` was a larger, quite nested struct) until I ran into an index out of range error which was quite annoying to debug at first.

My suggestion: check the errors and let them bubble up the stack.

(Here's the offending code: https://github.com/getsentry/raven-go/blob/dffeb57df75d6a911f00232155194e43d79d38d7/client.go#L245)

Scanning through the library, there's a fair amount of other places where errors are swallowed.

@mattrobenolt minimal repro: https://play.golang.org/p/7SxvL55pM_

Just swap in your own DSN and you'll get something like:

```
panic: runtime error: slice bounds out of range

goroutine 5 [running]:
panic(0x7d08c0, 0xc82000a0e0)
	/usr/lib/go/src/runtime/panic.go:481 +0x3e6
sermocrm.com/vendor/github.com/getsentry/raven-go.(*Packet).JSON(0xc8200bc600, 0x0, 0x0, 0x0)
	/home/eric/gopath/src/sermocrm.com/vendor/github.com/getsentry/raven-go/client.go:246 +0x3e4
sermocrm.com/vendor/github.com/getsentry/raven-go.serializedPacket(0xc8200bc600, 0x0, 0x0, 0x0, 0x0)
	/home/eric/gopath/src/sermocrm.com/vendor/github.com/getsentry/raven-go/client.go:684 +0x58
sermocrm.com/vendor/github.com/getsentry/raven-go.(*HTTPTransport).Send(0xc820015500, 0xc820015680, 0x2a, 0xc82007e580, 0x74, 0xc8200bc600, 0x0, 0x0)
	/home/eric/gopath/src/sermocrm.com/vendor/github.com/getsentry/raven-go/client.go:666 +0x6a
sermocrm.com/vendor/github.com/getsentry/raven-go.(*Client).worker(0xc8200be210)
	/home/eric/gopath/src/sermocrm.com/vendor/github.com/getsentry/raven-go/client.go:423 +0x166
created by sermocrm.com/vendor/github.com/getsentry/raven-go.newClient
	/home/eric/gopath/src/sermocrm.com/vendor/github.com/getsentry/raven-go/client.go:306 +0x193
exit status 2
```

--- a/client.go
+++ b/client.go
@@ -236,8 +236,11 @@ func uuid() (string, error) {
 	return hex.EncodeToString(id), nil
 }
 
-func (packet *Packet) JSON() []byte {
-	packetJSON, _ := json.Marshal(packet)
+func (packet *Packet) JSON() ([]byte, error) {
+	packetJSON, err := json.Marshal(packet)
+	if err != nil {
+		return nil, err
+	}
 
 	interfaces := make(map[string]Interface, len(packet.Interfaces))
 	for _, inter := range packet.Interfaces {
@@ -247,12 +250,15 @@ func (packet *Packet) JSON() []byte {
 	}
 
 	if len(interfaces) > 0 {
-		interfaceJSON, _ := json.Marshal(interfaces)
+		interfaceJSON, err := json.Marshal(interfaces)
+		if err != nil {
+			return nil, err
+		}
 		packetJSON[len(packetJSON)-1] = ','
 		packetJSON = append(packetJSON, interfaceJSON[1:]...)
 	}
 
-	return packetJSON
+	return packetJSON, nil
 }
 
 type context struct {
@@ -735,8 +741,14 @@ func (t *HTTPTransport) Send(url, authHeader string, packet *Packet) error {
 		return nil
 	}
 
-	body, contentType := serializedPacket(packet)
-	req, _ := http.NewRequest("POST", url, body)
+	body, contentType, err := serializedPacket(packet)
+	if err != nil {
+		return fmt.Errorf("error serializing packet: %v", err)
+	}
+	req, err := http.NewRequest("POST", url, body)
+	if err != nil {
+		return fmt.Errorf("can't create new request: %v", err)
+	}
 	req.Header.Set("X-Sentry-Auth", authHeader)
 	req.Header.Set("User-Agent", userAgent)
 	req.Header.Set("Content-Type", contentType)
@@ -752,8 +764,11 @@ func (t *HTTPTransport) Send(url, authHeader string, packet *Packet) error {
 	return nil
 }
 
-func serializedPacket(packet *Packet) (r io.Reader, contentType string) {
-	packetJSON := packet.JSON()
+func serializedPacket(packet *Packet) (io.Reader, string, error) {
+	packetJSON, err := packet.JSON()
+	if err != nil {
+		return nil, "", fmt.Errorf("error marshaling packet %+v to JSON: %v", packet, err)
+	}
 
 	// Only deflate/base64 the packet if it is bigger than 1KB, as there is
 	// overhead.
@@ -764,9 +779,9 @@ func serializedPacket(packet *Packet) (r io.Reader, contentType string) {
 		deflate.Write(packetJSON)
 		deflate.Close()
 		b64.Close()
-		return buf, "application/octet-stream"
+		return buf, "application/octet-stream", nil
 	}
-	return bytes.NewReader(packetJSON), "application/json"
+	return bytes.NewReader(packetJSON), "application/json", nil
 }
 
 var hostname string

--- a/client_test.go
+++ b/client_test.go
@@ -35,7 +35,11 @@ func TestPacketJSON(t *testing.T) {
 	packet.AddTags(map[string]string{"baz": "buzz"})
 
 	expected := `{"message":"test","event_id":"2","project":"1","timestamp":"2000-01-01T00:00:00.00","level":"error","logger":"com.getsentry.raven-go.logger-test-packet-json","platform":"linux","culprit":"caused_by","server_name":"host1","release":"721e41770371db95eee98ca2707686226b993eda","environment":"production","tags":[["foo","bar"],["foo","foo"],["baz","buzz"]],"modules":{"foo":"bar"},"fingerprint":["{{ default }}","a-custom-fingerprint"],"logentry":{"message":"foo"}}`
-	actual := string(packet.JSON())
+	j, err := packet.JSON()
+	if err != nil {
+		t.Fatalf("JSON marshalling should not fail: %v", err)
+	}
+	actual := string(j)
 
 	if actual != expected {
 		t.Errorf("incorrect json; got %s, want %s", actual, expected)
@@ -62,7 +66,11 @@ func TestPacketJSONNilInterface(t *testing.T) {
 	}
 
 	expected := `{"message":"test","event_id":"2","project":"1","timestamp":"2000-01-01T00:00:00.00","level":"error","logger":"com.getsentry.raven-go.logger-test-packet-json","platform":"linux","culprit":"caused_by","server_name":"host1","release":"721e41770371db95eee98ca2707686226b993eda","environment":"production","tags":[["foo","bar"]],"modules":{"foo":"bar"},"fingerprint":["{{ default }}","a-custom-fingerprint"],"logentry":{"message":"foo"}}`
-	actual := string(packet.JSON())
+	j, err := packet.JSON()
+	if err != nil {
+		t.Fatalf("JSON marshalling should not fail: %v", err)
+	}
+	actual := string(j)
 
 	if actual != expected {
 		t.Errorf("incorrect json; got %s, want %s", actual, expected)

